src/centroiders.cpp

```diff
@@ -1,45 +1,642 @@
+    return res;
+}
+
+struct FloodParams{
+    int xMin;
+    int xMax;
+    int yMin;
+    int yMax;
+};
+
+std::vector<Point> FloodfillPreproc(unsigned char *img, int imageWidth, int imageHeight) {
+    std::vector<Point> res;
+    std::set<Point> checkedPoints;
+
+    std::vector<unsigned char> image = CopyImage(img, imageWidth, imageHeight);
+
+    float noise;
+    int cutoff = BasicThreshold(image.data(), imageWidth, imageHeight, &noise);
+    cutoff -= noise;
+    SubtractNoise(image.data(), imageWidth, imageHeight, noise);
+
+    std::cout << "Cutoff: " << cutoff << std::endl;
+    if(cutoff == 0){
+        // TODO: scuffed, floodfill will never stop if cutoff=0, so stop it here
+        std::cerr << "No stars detected in image, killing process" << std::endl;
+        exit(EXIT_FAILURE);
+    }
+
+    for (long i = 0; i < imageHeight * imageWidth; i++) {
+        int x = i % imageWidth;
+        int y = i / imageWidth;
+        Point pCurr{x, y};
+        if (image[i] >= cutoff && checkedPoints.count(pCurr) == 0) {
+            // Floodfill from this point (x, y)
+            // Obtain coordinates (x0, y0) of pixel with largest magnitude in the floodfill
+            std::vector<Point> pts;
+            std::deque<Point> queue;
+
+            FloodParams fp{x, x, y, y};
+
+            int maxMag = -1;
+            int x0 = x;
+            int y0 = y;
+
+            int floodSize = 0;
+
+            queue.push_back(pCurr);
+            while (queue.size() != 0) {
+                Point p = queue[0];
+                queue.pop_front();
+
+                int px = p[0];
+                int py = p[1];
+                if (px < 0 || px >= imageWidth || py < 0 || py >= imageHeight) continue;
+                if (checkedPoints.count(p) != 0) continue;
+
+                int mag = Get(px, py, image.data(), imageWidth);
+                if (mag < cutoff) continue;
+
+                floodSize++;
+
+                fp.xMin = std::min(fp.xMin, px);
+                fp.xMax = std::max(fp.xMax, px);
+                fp.yMin = std::min(fp.yMin, py);
+                fp.yMax = std::max(fp.yMax, py);
+
+                // Add this point to pts and checkedPoints
+                // We can add to checkedPoints since cutoff is global - ensure no 2 fills collide
+                pts.push_back(p);
+                checkedPoints.insert(p);
+
+                // Update max pixel value in fill
+                if (mag > maxMag) {
+                    maxMag = mag;
+                    x0 = px;
+                    y0 = py;
+                }
+
+                // Add all 8 adjacent points to the queue
+                for (int dx = -1; dx <= 1; dx++) {
+                    for (int dy = -1; dy <= 1; dy++) {
+                        queue.push_back({px + dx, py + dy});
+                    }
+                }
+            }
+
+            // Cool, our flood is done
+            res.push_back({x0, y0, floodSize, fp.xMax - fp.xMin, fp.yMax - fp.yMin});
```

Comment:
`res` is a `std::vector<std::vector<int>>`. However, every element has the same structure, so there's no reason to use a vector -- it should be a vector of structs. It's okay to make single-use structs. (And while I like to use structs instead of tuples, a tuple is still better than a vector here).

(I'm going to fix this myself because I'm rewriting a whole bunch of centroid stuff right now, just thought you should know)